doc/man/man3/ds_list.3.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man3/ds_list.3.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH DS_LIST 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.\"
.\"
.SH NAME
.B ds_list
\- enumerate the immediate subordinates of a particular directory entry
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <xds.h>
.HP
.BI "DS_status ds_list(OM_private object " session ", OM_private_object " context ", OM_object " name ", OM_private_object *" result_return ", OM_sint *" invoke_id_return );
.\"
.\"
.SH INTERFACE
.PP
XDS
.[
XDS
.]
.\"
.\"
.SH DESCRIPTION
.PP
This function is used to obtain a list of the immediate subordinates of the
named entry.  The list may be incomplete in some circumstances.
.PP
The result of this operation can be abandonned.
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.IR session "\0\fBSession\fP(Object(Session)).
The directory session agains which this operation is performed.  This must be a
private object.
.TP
.IR context "\0\fBContext\fP(Object(Context)).
The directory context to be used for this operation.  This argument must be
a private object or the constant
.RB { DS_DEFAULT_CONTEXT }.
.TP
.IR name "\0\fBName\fP(Object(Name)).
The name of the object entry whose immediate subordinates are to be listed.
Any aliases in the name will be dereferenced unless prohibited by the relevant
service control.
.\"
.\"
.SH RETURN
.PP
.TP
.BR Status "(Status)
Takes the value
.B success
if the named object was located (even if there are no subordinates) and takes
an error value if not, when the subroutine is used synchronously.  Reports
whether the operation was initiated, if used asynchronously.
.TP
.BR Result "(Object(List-Result))
Upon successful completion of a synchronous call, the result contains some
information about the target object's immediate subordinates.  It also contains
the distinguished name of the target object, if an alias was dereferenced to
find it.  Aliases in the subordinate names are not dereferenced.  Also, there
may be a partial outcome qualifier, which indicates that the result is
incomplete.  It also explains why (for example, because the time limit expired)
and contains information that may be helpful when attempting to complete it.
.TP
.BR Invoke-ID "(Integer)
The
.B Invoke-ID
of an asynchronous directory operation.
.\"
.\"
.SH ERRORS
.PP
This function can return a System-Error or one of the following Library-Errors:
bad-argument, bad-class, bad-context, bad-name, bad-session, bad-workspace,
miscellaneous, missing-type, too-many-operations.
.PP
The following Directory-Errors may be returned by the function (or by
.BR ds_receive_result (3)
when used asyncrhonously):
Name-Error, Referral, Security-Error, Service-Error.
.PP
This function can return a Communications-Error.
.PP
This function can return the error constant
.RB [ DS_NO_WORKSPACE ].
.\"
.\"
.\".SH USAGE
.\".PP
.\"
.\"
.\".SH NOTICES
.\".PP
.\"
.\"
.SH IMPLEMENTATION
.PP
.BR ds_list ()
is implemented on top of the more complicated OSMIS 3.0 interface.
.\"
.\"
.\".SH FILES
.\".PP
.\"
.\"
.SH BUGS
.PP
.BR ds_list ()
has no known bugs.
.\"
.\"
.SH SEE ALSO
.PP
.BR XDS (3),
.BR XOM (3),
.BR ds_abandon (3),
.BR ds_receive_result (3).
.\"
.\"
.SH COMPATIBILITY
.PP
This implementation is compatible with other implementations of the XDS.
.[
XDS
.]
.\"
.\"
.SH CONFORMANCE
.PP
XDS.
.[
XDS
.]
Conformance is validated using an
.BR autotest (1)
script.
.\"and the
.\".BR test-xds (8)
.\"test suite executable.
.\"
.\"
.SH HISTORY
.PP
.BR ds_list ()
was first part of the
.BR XDS .
.[
XDS
.]
.\"
.\"
.XX
.[
xom c607
.]
.\"
.\"
.[
$LIST$
.]
.TA


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


isode/quipu/ds_list.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2007  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

/* ds_list.c - */

#ifndef lint
static char *rcsid =
    "Header: /xtel/isode/isode/quipu/RCS/ds_list.c,v 9.0 1992/06/16 12:34:01 isode Rel";
#endif

/*
 * Header: /xtel/isode/isode/quipu/RCS/ds_list.c,v 9.0 1992/06/16 12:34:01 isode Rel
 *
 *
 * Log: ds_list.c,v
 * Revision 9.0  1992/06/16  12:34:01  isode
 * Release 8.0
 *
 */

/*
 *                                NOTICE
 *
 *    Acquisition, use, and distribution of this module and related
 *    materials are subject to the restrictions of a license agreement.
 *    Consult the Preface in the User's Manual for the full terms of
 *    this agreement.
 *
 */

#include "quipu/util.h"
#include "quipu/connection.h"
#include "quipu/list.h"
#include "quipu/turbo.h"
#include "pepsy.h"
#include "quipu/DAS-types.h"

extern LLog *log_dsap;
extern Entry database_root;
static int build_result();

do_ds_list(arg, error, result, binddn, target, di_p, dsp, authtype)
	register struct ds_list_arg *arg;
	register struct ds_list_result *result;
	struct DSError *error;
	DN binddn;
	DN target;
	struct di_block **di_p;
	char dsp;
	char authtype;
{
	Entry entryptr;
	int retval;
	DN realtarget;
	int sizelimit;
	int authp;

	DLOG(log_dsap, LLOG_TRACE, ("ds_list"));

	if (!dsp)
		target = arg->lsa_object;

	switch (find_child_entry
		(target, &(arg->lsa_common), binddn, NULLDNSEQ, FALSE, &(entryptr), error, di_p)) {
	case DS_OK:
		/* Filled out entryptr - carry on */
		break;
	case DS_CONTINUE:
		/* Filled out di_p - what do we do with it ?? */
		return (DS_CONTINUE);

	case DS_X500_ERROR:
		/* Filled out error - what do we do with it ?? */
		return (DS_X500_ERROR);
	default:
		/* SCREAM */
		LLOG(log_dsap, LLOG_EXCEPTIONS, ("do_ds_list() - find_child_entry failed"));
		return (DS_ERROR_LOCAL);
	}

	/* Strong authentication */
	if ((retval = check_security_parms((caddr_t) arg,
					   _ZListArgumentDataDAS,
					   &_ZDAS_mod,
					   arg->lsa_common.ca_security,
					   arg->lsa_common.ca_sig, &binddn)) != 0) {
		error->dse_type = DSE_SECURITYERROR;
		error->ERR_SECURITY.DSE_sc_problem = retval;
		return (DS_ERROR_REMOTE);
	}

	realtarget = get_copy_dn(entryptr);

	if (isleaf(entryptr)) {

		dn_free(realtarget);

		result->lsr_subordinates = NULLSUBORD;
		result->lsr_age = (time_t) 0;
		result->lsr_common.cr_requestor = NULLDN;
		if (error->dse_type == DSE_NOERROR) {
			result->lsr_object = NULLDN;
			result->lsr_common.cr_aliasdereferenced = FALSE;
		} else {
			result->lsr_common.cr_aliasdereferenced = TRUE;
			result->lsr_object = get_copy_dn(entryptr->e_parent);
		}
		result->lsr_cr = NULLCONTINUATIONREF;
		result->lsr_limitproblem = LSR_NOLIMITPROBLEM;
		return (DS_OK);
	}

	/* do authentication policy stuff here */
	if (!manager(binddn))
		authp = entryptr->e_authp ? entryptr->e_authp->ap_listandsearch : AP_SIMPLE;
	else
		authp = AP_SIMPLE;

	/* check parent will allow listing */
	sizelimit = SVC_NOSIZELIMIT;
	if (entryptr->e_lacl != NULLAV) {
		if (check_lacl((authtype % 3) >= authp ? binddn : NULLDN,
			       realtarget, entryptr->e_lacl, SACL_SINGLELEVEL,
			       &sizelimit) == NOTOK) {
			/* security error */
			error->dse_type = DSE_SECURITYERROR;
			error->ERR_SECURITY.DSE_sc_problem = DSE_SC_ACCESSRIGHTS;
			dn_free(realtarget);
			return (DS_ERROR_REMOTE);
		}
	}
	if (check_acl((authtype % 3) >= authp ? binddn : NULLDN,
		      ACL_READ, entryptr->e_acl->ac_child, realtarget) != OK) {
		if (dsp
		    && (check_acl(binddn, ACL_READ, entryptr->e_acl->ac_child, realtarget) == OK)) {
			error->dse_type = DSE_SECURITYERROR;
			error->ERR_SECURITY.DSE_sc_problem = DSE_SC_AUTHENTICATION;
			dn_free(realtarget);
			return (DS_ERROR_REMOTE);
		}
		error->dse_type = DSE_SECURITYERROR;
		error->ERR_SECURITY.DSE_sc_problem = DSE_SC_ACCESSRIGHTS;
		dn_free(realtarget);
		return (DS_ERROR_REMOTE);
	}

	if (entryptr->e_children == NULLAVL) {
		int res;

		if (try_cache(arg, result, realtarget) == OK) {
			dn_free(realtarget);
			return (DS_OK);
		}
		res = constructor_dsa_info(realtarget, NULLDNSEQ, FALSE, entryptr, error, di_p);
		dn_free(realtarget);
		return res;
	}

	dn_free(realtarget);

	build_result(arg, entryptr, result, error,
		     (authtype % 3) >= authp ? binddn : NULLDN, dsp, sizelimit);
	return (DS_OK);
}

/*
 * these are globals changed by build_result and build_list.  they have
 * to be globals because build_list is called by avl_apply, and it only
 * allows one additional argument to be passed.  pretty gross, but that's
 * the way it is...
 */

static struct subordinate *g_sub;
static struct subordinate *g_trail = NULLSUBORD;
static int g_count;
static int g_size;
static int g_dsp;
static int g_security;
static DN g_dn;
static DN g_dnend;

static int
build_list(e, dn)
	Entry e;
	DN dn;
{
	struct subordinate *sub;

	if (g_size != SVC_NOSIZELIMIT && g_count >= g_size)
		return (NOTOK);

	g_dnend->dn_rdn = e->e_name;

	if (g_dsp) {
		/* Only send publicly read data over DSP */

		if (e->e_lacl != NULLAV) {
			int dummy;

			if (check_lacl(NULLDN, g_dn, e->e_lacl, SACL_BASEOBJECT, &dummy) != OK) {

				if (check_lacl(dn, g_dn, e->e_lacl, SACL_BASEOBJECT, &dummy) == OK)
					/* Would get more over DAP */
					g_security = 1;

				return (0);
			}
		} else {
			if (check_acl(NULLDN, ACL_READ, e->e_acl->ac_entry, g_dn) != OK) {
				if (check_acl(dn, ACL_READ, e->e_acl->ac_entry, g_dn) != OK)
					/* Would get more over DAP */
					g_security = 1;
				return (0);
			}
		}

	} else {

		if (e->e_lacl != NULLAV) {
			int dummy;

			if (check_lacl(dn, g_dn, e->e_lacl, SACL_BASEOBJECT, &dummy) != OK)
				return (0);

		} else if (check_acl(dn, ACL_READ, e->e_acl->ac_entry, g_dn) != OK)
			return (0);

	}
	sub = (struct subordinate *) smalloc(sizeof(struct subordinate));
	sub->sub_copy = e->e_data;
	sub->sub_rdn = rdn_cpy(e->e_name);
	sub->sub_aliasentry = (e->e_alias == NULLDN ? FALSE : TRUE);
	sub->sub_next = NULLSUBORD;

	if (g_sub == NULLSUBORD) {
		g_sub = sub;
		g_trail = sub;
	} else {
		g_trail->sub_next = sub;
		g_trail = sub;
	}

	g_count++;
	return (0);
}

static int
build_result(arg, ptr, result, error, binddn, dsp, laclsizelimit)
	register Entry ptr;
	struct ds_list_arg *arg;
	struct ds_list_result *result;
	struct DSError *error;
	DN binddn;
	char dsp;
	int laclsizelimit;
{
	DN dn;
	DN dnend;
	int size;
	RDN dnrdn;
	extern int admin_size;
	char adminlimit = FALSE;
	int rc;
	Entry akid;

	DLOG(log_dsap, LLOG_DEBUG, ("building list results"));

	result->lsr_subordinates = NULLSUBORD;
	if (!dsp && manager(binddn))
		size = arg->lsa_common.ca_servicecontrol.svc_sizelimit;
	else {
		if (laclsizelimit == SVC_NOSIZELIMIT)
			laclsizelimit = admin_size;

		if ((size = MIN(laclsizelimit, MIN(admin_size,
						   arg->lsa_common.ca_servicecontrol.
						   svc_sizelimit)))
		    == SVC_NOSIZELIMIT) {
			size = MIN(laclsizelimit, admin_size);
			adminlimit = TRUE;
		}
	}

	result->lsr_age = (time_t) 0;
	result->lsr_common.cr_requestor = NULLDN;
	/* if no error and NOT SVC_OPT_DONTDEREFERENCEALIASES then */
	/* the alias will have been derefeferenced -signified by */
	/* NO_ERROR !!! */
	if (error->dse_type == DSE_NOERROR) {
		result->lsr_object = NULLDN;
		result->lsr_common.cr_aliasdereferenced = FALSE;
	} else {
		result->lsr_common.cr_aliasdereferenced = TRUE;
		result->lsr_object = get_copy_dn(ptr);
	}
	result->lsr_cr = NULLCONTINUATIONREF;

	/* we already checked for null kids ... */
	akid = (Entry) avl_getone(ptr->e_children);
	dn = get_copy_dn(akid);
	for (dnend = dn; dnend->dn_parent != NULLDN; dnend = dnend->dn_parent) ;	/* NO-OP */
	dnrdn = dnend->dn_rdn;

	g_dn = dn;
	g_dsp = dsp;
	g_dnend = dnend;
	g_sub = NULLSUBORD;
	g_size = size;
	g_count = 0;
	g_security = 0;

	/* 
	 * preorder would be a little faster in case of small size limit,
	 * but inorder is more user-predictable, which is nice, though not
	 * required...
	 */

	rc = avl_apply(ptr->e_children, build_list, (caddr_t) binddn, NOTOK, AVL_INORDER);

	/* 
	 * build_list has updated g_count and g_sub to contain a count of
	 * the number of entries in the list and the list itself,
	 * respectively.  if avl_apply was not cut short because the size
	 * limit was reached (i.e. instead it ran out of nodes), rc will
	 * be NOTOK.
	 */

	size = g_size;
	result->lsr_subordinates = g_sub;

	if (rc != AVL_NOMORE)
		/* stopped look up due to size limit */
		/* need to send continuation reference */
		result->lsr_limitproblem = adminlimit ?
		    LSR_ADMINSIZEEXCEEDED : LSR_SIZELIMITEXCEEDED;
	else
		result->lsr_limitproblem = LSR_NOLIMITPROBLEM;

	if (g_security)
		/* Could force a DAP operation - this is easier ! */
		result->lsr_limitproblem = LSR_ADMINSIZEEXCEEDED;

	dnend->dn_rdn = NULLRDN;
	dn_free(dn);
	rdn_free(dnrdn);
}

try_cache(arg, result, target)
	register struct ds_list_arg *arg;
	register struct ds_list_result *result;
	DN target;
{
	struct list_cache *ptr;
	struct subordinate *subord_cpy();

	if ((arg->lsa_common.ca_servicecontrol.svc_options & SVC_OPT_DONTUSECOPY) == 0) {
		if ((ptr =
		     find_list_cache(target,
				     arg->lsa_common.ca_servicecontrol.svc_sizelimit)) !=
		    NULLCACHE) {
			DLOG(log_dsap, LLOG_DEBUG, ("building list results using cache"));
			result->lsr_subordinates = subord_cpy(ptr->list_subs);
			result->lsr_age = (time_t) 0;
			result->lsr_common.cr_aliasdereferenced = FALSE;
			result->lsr_common.cr_requestor = NULLDN;
			result->lsr_object = NULLDN;
			result->lsr_cr = NULLCONTINUATIONREF;
			result->lsr_limitproblem = ptr->list_problem;
			return (OK);
		}
	}

	return (NOTOK);
}
